### PR TITLE
Add ability to preload entity script whitelist

### PR DESCRIPTION
**Pull request by digisomni**
I also renamed the "INITIAL_STARTUP_LOCATION" to "PRELOADED_STARTUP_LOCATION".

To test this feature, you have to have the environment variable set when CMake is run. The whitelist should then be activated and also be prepopulated with the values that you had set.

I intend to move all script whitelist/blocking/security into this .cpp in the future, actually. But for now I've just gone ahead and made it.

### CODE REVIEW

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1082/Vircadia-PR1082-4f972ef-.tar_b5d89c44.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1082/Vircadia-PR1082-4f972ef-.tar_b5d89c44.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1082/Vircadia-PR1082-4f972ef-_b5d89c44.sh

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1082/Vircadia-PR1082-4f972ef-_b5d89c44.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1082/Vircadia-PR1082-4f972ef-_b5d89c44.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1082/Vircadia-PR1082-4f972ef_b5d89c44.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1082/Vircadia-Quest-Alpha-PR1082-4f972ef_b5d89c44.apk

File: libraries/script-engine/src/ScriptGatekeeper.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+#include "ScriptGatekeeper.h"
+
+#include "BuildInfo.h"
+#include "SettingHandle.h"
+
+void ScriptGatekeeper::initialize() {
+    if (_initialized == true) {
```
Comment (ctrlaltdavid) [inline libraries/script-engine/src/ScriptGatekeeper.cpp:18]:
```suggestion
    if (_initialized) {
```

File: libraries/script-engine/src/ScriptGatekeeper.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        return;
+    }
+
+    QVariant rawCurrentWhitelistValues = Setting::Handle<QVariant>(SCRIPT_WHITELIST_ENTRIES_KEY).get();
+    QString settingsSafeValues = rawCurrentWhitelistValues.toString();
+
+    Setting::Handle<bool> whitelistEnabled{ SCRIPT_WHITELIST_ENABLED_KEY, false };
```
Comment (ctrlaltdavid) [inline libraries/script-engine/src/ScriptGatekeeper.cpp:25]:
```suggestion
    Setting::Handle<bool> whitelistEnabled { SCRIPT_WHITELIST_ENABLED_KEY, false };
```

File: libraries/script-engine/src/ScriptGatekeeper.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    Setting::Handle<bool> isFirstRun { Settings::firstRun, true };
+
+    QString preloadedVal = BuildInfo::PRELOADED_SCRIPT_WHITELIST;
+
+    if (settingsSafeValues.isEmpty() && !preloadedVal.isEmpty() && isFirstRun.get()) {
+        // We assume that the whitelist should be enabled if a preloaded whitelist is attached, so we activate it if it's not already active.
+        qDebug() << "hi" << !whitelistEnabled.get() << whitelistEnabled.get();
```
Comment (ctrlaltdavid) [inline libraries/script-engine/src/ScriptGatekeeper.cpp:32]: Remove debug?

File: libraries/script-engine/src/ScriptGatekeeper.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+class ScriptGatekeeper : public QObject {
+    Q_OBJECT
+public:
+    void initialize();
+
+    QString SCRIPT_WHITELIST_ENABLED_KEY{ "private/whitelistEnabled" };
+    QString SCRIPT_WHITELIST_ENTRIES_KEY{ "private/settingsSafeURLS" };
+
+private:
+    bool _initialized{ false };
```
Comment (ctrlaltdavid) [inline libraries/script-engine/src/ScriptGatekeeper.h:26]:
```suggestion
    bool _initialized { false };
```

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

**Review by ctrlaltdavid — APPROVED** [review]

**Comment by Aitolda** [comment]
Seems to be working per verbal instructions by Kalila.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1082/Vircadia-PR1082-72dda00-.tar_d8a3d56d.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1082/Vircadia-PR1082-72dda00-.tar_d8a3d56d.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1082/Vircadia-PR1082-72dda00-_d8a3d56d.sh

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1082/Vircadia-PR1082-72dda00-_d8a3d56d.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1082/Vircadia-PR1082-72dda00-_d8a3d56d.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1082/Vircadia-Quest-Alpha-PR1082-72dda00_d8a3d56d.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1082/Vircadia-PR1082-72dda00_d8a3d56d.apk

**Pull request merged**